### DISCUSSION TITLE
[@types/p5] Types are lacking generic overloads

### DISCUSSION THREAD

**Discussion by EndBug** [DefinitelyTyped/DefinitelyTyped#61291] [Issues with a @types package] [answered]
Some methods have overloads, but the generic one is missing. Here's an example:

```ts
p5.stroke('string') // OK
p5.stroke(p5.color(255, 204, 0)) // OK

let a: string | p5.Color
p5.stroke(a)
/*
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'string | Color' is not assignable to parameter of type 'Color'.
      Type 'string' is not assignable to type 'Color'.
*/
```

Reference: https://www.typescriptlang.org/docs/handbook/2/functions.html#writing-good-overloads

[a comment by an automated account is not shown]

**Comment by Gaweph** [comment]
Fur this specific example, it looks like the are actually a few overloads? E.g `number[]` `number, number` etc...

Would a single overload that looks something like `string|p5.Color|number[]` be the answer? Or would we have to create one for each combination of overloads?

For info to help you understand where these come from, we are using the `jsdoc` comments in the main `p5` solution to generate the TS definition.

Here's the `sketch` method if you are interested in having a look.

https://github.com/processing/p5.js/blob/e32b45367baad694b1f4eeec0586b910bfcf0724/src/color/setting.js#L748

**Comment by EndBug** [comment, marked as answer]
Tbh looking into it seems like a bit of a mess: a generic type would solve the issue, but at the same time it would make the results from type inference more generic, and also would make it more difficult to get JSDoc comments in the editor...
It also seems that TypeScript itself compiles declarations of functions like this without the generic overload ([playgound](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMAFAQwFyLCAtgIwFMAnAGkXwH4MBnKYmMAcwEoscCSBYAKFElgJk6LHQbNyVDByLE2iMYya9+0eElSZseWQB9FEitW2di++kpaIA3r0SJihKCGJI0vAL68VKAORpfRDQaEz0DViA)) so I guess it's just something we'll have to deal with?

Unless you have a better idea on how to handle it, I'll close the discussion. Thanks either way ❤️
